### PR TITLE
METRON-1121 Ignore Profile with Bad 'init', 'update' or 'groupBy'

### PR DESCRIPTION
**Pull request by nickwallen**
The Profiler needs to gracefully ignore profiles that have a bad 'init', 'update', or 'groupBy' expression.  Under certain conditions a message may not be applied to all valid profiles, due to a single bad profile.

If the Profiler is running with multiple profiles defined and one of those profiles has a Stellar expression that throws an exception for either the  'init', 'update', or 'groupBy', the current logic will not continue to attempt to apply the message to the remaining valid profiles.  Only valid profiles executed before the bad profile will have the message applied.

## Testing

Unit tests have been added to check for these conditions.  This can be manually tested with either a live Profiler running in a cluster or using the Profiler debugging functions.  

Create two profiles; one that is valid and another that is invalid, like the following.  Ensure that all test messages that are sent through the Profiler, reach the `good-profile` despite the `bad-profile`.

```
 {
   "profiles": [
      {
        "profile": "bad-profile",
        "foreach": "ip_src_addr",
        "init":   { "x": "2 / 0" },
        "update": { "x": "x + 1" },
        "result": "x"
      },
      {
        "profile": "good-profile",
        "foreach": "ip_src_addr",
        "init":   { "x": "0" },
        "update": { "x": "x + 1" },
        "result": "x"
      }
   ]
 }
```

## Pull Request Checklist

- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

### CODE REVIEW

**Comment by cestella** [comment]
+1 by inspection

File: metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java [hunk 1]
```diff
@@ -140,33 +144,41 @@ public void apply(JSONObject message) {
...
+      result = Optional.of(new ProfileMeasurement()
+              .withProfileName(profileName)
+              .withEntity(entity)
+              .withGroups(groups)
+              .withPeriod(clock.currentTimeMillis(), periodDurationMillis, TimeUnit.MILLISECONDS)
+              .withProfileValue(profileValue)
+              .withTriageValues(triageValues)
+              .withDefinition(definition));
+
+    } catch(Exception e) {
```
Comment (cestella) [inline metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java]: Should this be a Throwable?

File: metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java [hunk 2]
```diff
@@ -140,33 +144,41 @@ public void apply(JSONObject message) {
...
+      // if any of the Stellar expressions fail, a measurement should NOT be returned
+      LOG.error(format("Unable to flush profile: %s", e.getMessage()));
```
Comment (cestella) [inline metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java]: Can we pass the exception as the second argument to `LOG.error`

File: metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java [hunk 3]
```diff
@@ -253,7 +265,8 @@ private void assign(Map<String, String> expressions, Map<String, Object> transie
               .forEach((expr) -> results.add(executor.execute(expr, transientState, Object.class)));
 
     } catch (Throwable e) {
-      String msg = format("Bad '%s' expression: %s, profile=%s, entity=%s", expressionType, e.getMessage(), profileName, entity);
+      String msg = format("Bad '%s' expression: error=%s, profile=%s, entity=%s", expressionType, e.getMessage(), profileName, entity);
+      LOG.error(msg);
```
Comment (cestella) [inline metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/DefaultProfileBuilder.java]: Can you pass `e` as the second argument to `LOG.error`?

**Comment by nickwallen** [comment]
Good eye, @cestella .  Latest commit is catching Throwable and logging e's.

**Comment by cestella** [comment]
Yep, +1 by inspection, thanks!

**Pull request closed without merging**